Add description about Gen-AI contributions to our guide (#60158)
mistercrunch/Airflow: .github/PULL_REQUEST_TEMPLATE.md

<!--
Thank you for contributing!

Please provide above a brief description of the changes made in this pull request.
Write a good git commit message following this guide: http://chris.beams.io/posts/git-commit/

Please make sure that your code changes are covered with tests.
And in case of new features or big changes remember to adjust the documentation.

Feel free to ping (in general) for the review if you do not see reaction for a few days
(72 Hours is the minimum reaction time you can expect from volunteers) - we sometimes miss notifications.

In case of an existing issue, reference it using one of the following:

* closes: #ISSUE
* related: #ISSUE
-->

## Was generative AI tooling used to co-author this PR?

<!--
If generative AI tooling has been used in the process of authoring this PR, please
change below checkbox to `[X]` followed by the name of the tool.
-->

- [ ] Yes (please specify the tool below)

> Generated-by: [Tool Name] following [the guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)

Reviewer: Why are we concerned with what the tool is? Is that critical information or more of a frustrating data entry for those of us who regularly commit?

Also is it just the tool that we want to know about? Or the model itself? I would think the model is more relevant?

I think something simpler like this would be better:

```
<!--
If generative AI tooling has been used in the process of authoring this PR, please
change the below checkbox to `[X]` to confirm the usage followed the [Gen AI project guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)
-->

- [ ] Yes, Gen AI was used and I confirm that the usage followed project guidelines.
```

Author: That's a good point and I already removed version (copied it from Spark template).

This is mostly about potential legal repercussions. And it's taken directly from:

https://www.apache.org/legal/generative-tooling.html

> When providing contributions authored using generative AI tooling, a recommended practice is for contributors to indicate the tooling used to create the contribution. This should be included as a token in the source control commit message, for example including the phrase “Generated-by: ”. This allows for future release tooling to be considered that pulls this content into a machine parsable Tooling-Provenance file.

So we are strictly following Guidelines of the ASF.

And yes - it is ambiguous and left to discretion of people what they provide there, and I think that's ok. Yes, it adds a bit of friction, but also I think some level of reasonable friction is open-source is actually a very good thing.  This will make some people pause and think what they used - and maybe even double check what their configuration is, and maybe even they will start asking questions about their tooling - which might be actually better than hiding that they are using something.

There is a great talk about dangers to open-source which I watched recently - on how sustainability of OSS is endangered now, where one of the reasons is that our tooling is "too easy" for people to participate and that creates a lot of troubles for maintainers, and adding some reasonable friction which make people start thinking and having to make choices, is a good thing. I highly recommend that one - https://media.ccc.de/v/39c3-how-to-keep-open-source-open-without-leaving-our-communities-open-to-threats - in fact I am signing up to the working group mentioned there to discuss and probably come up with some blueprints and patterns in the open-source on how to keep OSS sustainable.

Author: One problem though I have here indeed is that the "Generated-by" is always there - no matter if you check the box or not, which is a bit annoying, so I  might change it into a comment and ask people to uncomment it if they mark `X`

Reviewer: If it's ASF policy then I suppose we have no choice but to follow it (but we also weren't asking previously).
I do, however, think that it's a strange policy and in the months and years to come I reckon it will fall away. It will soon be like asking someone what IDE they used to develop with. As AI becomes more entrenched it will not matter much at all which tool you used (and I think it's a stretch even that it matters now).

Author: It's not a policy - it's a guideline - and we do not have to follow it, It's more of "would be nice to have it".

I changed it now into a commented entry - so it will not show up, and we only tell it "should" be done. But I guess a number of people will not follow even if they mark `[X]`. And that's fine as well. I think this is also an interesting signal to track (even if we are not formally making statistics now). More like experimenting with it to see what **really** happens.

I think friction is non-zero here, but also not huge (i.e. reasonable). And we need "a little reasonable friction here" I think.

Author: And yes - even tthe guidelines mention that they are likely to change, and there were already a few discussions in the ASF that might eventually lead to some changes there.

Author: Also one more comment here - ASF "policies" and "guidelines" are very well written - they only prescribe what is really a MUST as a MUST, and they leave a lot of freedom to both PMC and contributors to a) follow them b) understand the spirit c) or not.

The fact that we are not forcing people to do stuff - but we are telling them what they should do (and make it with  just a little friction) - is exactly intentional. Because by the sheer fact of either following them, or not, you are showing if you are good citizen or not. If we see people - clearly submitting AI genearated stuff but not checking the checkbox - that will be an immediate sign of being careful with the PR. Simple signal, but telling a lot. 

And if we see someone not only checking the box but also uncommenting the generated-by and filling it in it might mean one of the two things:

* good citizen user who is good prospect to watch and encourage
* very good AI that read the guidelnes and fulfilled them - which makes it likely they also fullfilled other expectations

Of course I am simplyfying it a bit, people will attempt to game it as well, but this at least gives us some better signal on the PR quality.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.

>  -->

Reviewer 2: Why remove this? Or another qn, why is/was the license needed?

Author: Good question. Thanks ! 

Because when you create PR - this one is present in the proposed text. And As our template gets longer - it was really adding a lot of noise.

We had it because our prek hook adds it automatically to all files. And we are supposed to add it to **any** file we are releasing as part of the "source" package, unless the file is generated. 

However, this file is not included in "sources" release. This line in .gitattributues means that we do not include content of `.github` when we run `git archive` to prepare the source tar-ball package.

```
.github export-ignore
```

So - we are save to remove the licence header because we are not releasing it.

> Feel free to ping (in general) for the review if you do not see reaction for a few days

Reviewer 2: ```suggestion
Feel free to ping (in general) for the review if you do not see reviews for a few days
```

Author: I think reaction is better - I put it deliberately here. Because the reaction might be "close" for example or "need resolve conflict", or "need more information" or something - without actually doing review (though of course one might say any reaction is kind of review ;)) .

Reviewer 2: Sure sound good

> ## Was generative AI tooling used to co-author this PR?

Reviewer 2: ```suggestion
## Was generative AI tooling used to support changes in this PR?
```

`co-author` might seem too specific? There could be cases when a contributor thinks, I used it in "ask" mode just for some tests, so thats likely not a co-author situation

Author: Actually - that was also quite deliberate. "Authoring" has a more broad meaning than "agentically adding the code" - even if you "ask" and use the suggestions of AI - AI is "co-author". Same as when in GitHub UI you suggested above, I might choose to commit your suggestion, which will make you co-author of this commit (and actually even adding automatically co-author in the commit message). 

Also this is the term used in https://www.apache.org/legal/generative-tooling.html  where "authoring using generative AI tooling" is used.

> When providing contributions authored using generative AI tooling, a recommended practice is for contributors to indicate the tooling used to create the contribution. This should be included as a token in the source control commit message, for example including the phrase “Generated-by: ”. This allows for future release tooling to be considered that pulls this content into a machine parsable Tooling-Provenance file.

A bit of philosophical context (It's a very interesting subject and I am following and participating in discussions about it in ASF):

I actually very much like "authoring" term - like for example Dag Authoring. This applies very broadly - anything that ends up in created code (in this case). For example using automated linting to create Dag is also somewhat "authoring" it with those tools. 

The big difference (and that also relates to @o-nikolas suggestion that we might be asked about IDEs) is that currently we are in a little bit of a legal vacuum - automated linting and IDE classic autocomplete (before AI) - is quite clearly non-copyrightable. But there is a debate going on if the AI generated code is copryightable. And there is no clear answer. Some legal precedents already exist and courts generally dismiss claims that AI generated code (and music, and images and videos etc.) is copyrightable according to the current "copyright" law. 

This is something we deeply discuss in the ASF, but I think it's a bit early to have a definite answer, and my thinking - and a number of people in ASF think the same - that we (industry and humanity) will have to come up with a different kinds of laws for this - because "copyright" does not apply to it at all. We have to remember that copyright law is rather new (for humanity [1]) and is really a "social agreement" invented to protect creators and when it was coined, anything like Gen AI was not even imaginable.

So "authoring" is a way to say - it was used to create it. And the only reason we expect it here (and why it was written like that in the guidelines) is precisely because the legal situation is not clear and we want to keep track of it just in case.

[1] https://en.wikipedia.org/wiki/Copyright

Reviewer 2: Thanks for clarification, that was quite a read but makes sense

Reviewer 2: Thanks for clarification, that was quite a read but makes sense

<!--
If generative AI tooling has been used in the process of authoring this PR, please
change below checkbox to `[X]` followed by the name of the tool, uncomment the "Generated-by".
-->

> - [ ] Yes (please specify the tool below)

Reviewer 2: ```suggestion
- [ ] Yes (please specify the tool below)
- [ ] No

```

We need a NO as well?

Author: I don't think so. Absence of marking Yes means NO. And my goal is that only people who use Gen AI are supposed  to do anything. If you created the PR without Gen AI, you should be able to do `git commit`, add description, push, create PR - without having to additionally type anything. I wanted **this** workflow to be as frictionless as possible.
